File: DevExpress-Demo/CS/SchedulingDemo.Wpf/ViewModels/Customization/TimeRegionsDemoViewModel.cs
using DevExpress.Mvvm.Native;
using DevExpress.Mvvm.POCO;
using DevExpress.Xpf.Scheduling;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace SchedulingDemo.ViewModels {
    public class TimeRegionsDemoViewModel {        
        public const string AnyResource = "(Any)";

        readonly TimeSpan regionDuration = TimeSpan.FromHours(3);        

        DateTime regionStart;
        int timeRegionNumber = 0;

        public virtual DateTime Start { get; set; }
        public IEnumerable<WorkCalendar> Calendars { get { return WorkData.Calendars; } }
        public IEnumerable<WorkAppointment> Appointments { get { return WorkData.Appointments; } }
        public virtual ObservableCollection<TimeRegionViewModel> TimeRegions { get; protected set; }
        public virtual TimeRegionViewModel SelectedTimeRegion { get; set; }
        public virtual IList<string> Resources { get; protected set; }

        public TimeRegionsDemoViewModel() {
            Start = WorkData.TodayStart;
            TimeRegions = new ObservableCollection<TimeRegionViewModel>();
            regionStart = Start.AddHours(9);

            Init();
        }
        void Init() {            
            WorkData.Calendars.ToList()
                .ForEach(x => x.IsVisible = false);
            WorkData.CalendarConferenceRoom.IsVisible = true;
            WorkData.CalendarTrainingRoom.IsVisible = true;
            Resources = WorkData.CalendarsRooms.Select(x => x.Name).ToList();
            Resources.Insert(0, AnyResource);

            DateTime customRegionStart = Start.AddDays(1).AddHours(9);
            AddCustomTimeRegion("Predefined TimeRegion1", customRegionStart, customRegionStart.AddMinutes(45), DefaultBrushNames.TimeRegion1Hatch);
            customRegionStart = Start.AddDays(1).AddHours(13);
            AddCustomTimeRegion("Predefined TimeRegion2", customRegionStart, customRegionStart.AddMinutes(15), DefaultBrushNames.TimeRegion8Dotted);
            customRegionStart = Start.AddDays(2).AddHours(11);
            AddCustomTimeRegion("Predefined TimeRegion3", customRegionStart, customRegionStart.AddMinutes(75), DefaultBrushNames.TimeRegion2Hatch, new long[] { 1 });
            customRegionStart = Start.AddDays(3).AddHours(10);
            AddCustomTimeRegion("Predefined TimeRegion4", customRegionStart, customRegionStart.AddHours(5), DefaultBrushNames.TimeRegion3Hatch);
            SelectedTimeRegion = TimeRegions[0];
        }

        void AddCustomTimeRegion(string caption, DateTime start, DateTime end, string brushName, IEnumerable<long> resourceIds = null) {
            var res = TimeRegionViewModel.Create();
            res.Caption = caption;
            res.Start = start;
            res.End = end;
            res.BrushName = brushName;
            res.ResourceIds = resourceIds;
            TimeRegions.Add(res);
        }
        void AddNewTimeRegionInternal(string caption) {
            var res = TimeRegionViewModel.Create();
            res.Caption = caption;
            res.Start = regionStart;
            res.End = regionStart + regionDuration;
            res.BrushName = DefaultBrushNames.TimeRegions[timeRegionNumber % DefaultBrushNames.TimeRegions.Length];
            TimeRegions.Add(res);

            timeRegionNumber++;
            regionStart += regionDuration;

            SelectedTimeRegion = res;
        }
        public void AddNewTimeRegion(string caption = null) {
            if (String.IsNullOrEmpty(caption))
                caption = "TimeRegion";
            AddNewTimeRegionInternal(caption + this.timeRegionNumber);
        }
        public void RemoveSelectedTimeRegion() {
            TimeRegions.Remove(SelectedTimeRegion);
            SelectedTimeRegion = TimeRegions.FirstOrDefault();
        }
    }

    public class TimeRegionViewModel {
        public static TimeRegionViewModel Create() {
            return ViewModelSource.Create(() => new TimeRegionViewModel());
        }
        protected TimeRegionViewModel() { }
        public virtual string Caption { get; set; }
        public virtual DateTime Start { get; set; }
        public virtual DateTime End { get; set; }
        public virtual string BrushName { get; set; }
        public virtual string BorderBrushName { get { return DefaultBrushNames.Resources[DefaultBrushNames.TimeRegions.IndexOf(x => x == BrushName) % DefaultBrushNames.Resources.Length]; } }
        public virtual IEnumerable<long> ResourceIds { get; set; }
    }
}

File: DevExpress-Demo/VB/SchedulingDemo.Wpf/ViewModels/Customization/TimeRegionsDemoViewModel.vb
Imports DevExpress.Mvvm.Native
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Xpf.Scheduling
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Linq

Namespace SchedulingDemo.ViewModels

    Public Class TimeRegionsDemoViewModel

        Public Const AnyResource As String = "(Any)"

        Private ReadOnly regionDuration As TimeSpan = TimeSpan.FromHours(3)

        Private regionStart As Date

        Private timeRegionNumber As Integer = 0

        Public Overridable Property Start As Date

        Public ReadOnly Property Calendars As IEnumerable(Of WorkCalendar)
            Get
                Return WorkData.Calendars
            End Get
        End Property

        Public ReadOnly Property Appointments As IEnumerable(Of WorkAppointment)
            Get
                Return WorkData.Appointments
            End Get
        End Property

        Public Overridable Property TimeRegions As ObservableCollection(Of TimeRegionViewModel)

        Public Overridable Property SelectedTimeRegion As TimeRegionViewModel

        Public Overridable Property Resources As IList(Of String)

        Public Sub New()
            Start = TodayStart
            TimeRegions = New ObservableCollection(Of TimeRegionViewModel)()
            regionStart = Start.AddHours(9)
            Init()
        End Sub

        Private Sub Init()
            Enumerable.ToList(WorkData.Calendars).ForEach(Sub(x) x.IsVisible = False)
            CalendarConferenceRoom.IsVisible = True
            CalendarTrainingRoom.IsVisible = True
            Resources = CalendarsRooms.[Select](Function(x) x.Name).ToList()
            Resources.Insert(0, AnyResource)
            Dim customRegionStart As Date = Start.AddDays(1).AddHours(9)
            AddCustomTimeRegion("Predefined TimeRegion1", customRegionStart, customRegionStart.AddMinutes(45), DefaultBrushNames.TimeRegion1Hatch)
            customRegionStart = Start.AddDays(1).AddHours(13)
            AddCustomTimeRegion("Predefined TimeRegion2", customRegionStart, customRegionStart.AddMinutes(15), DefaultBrushNames.TimeRegion8Dotted)
            customRegionStart = Start.AddDays(2).AddHours(11)
            AddCustomTimeRegion("Predefined TimeRegion3", customRegionStart, customRegionStart.AddMinutes(75), DefaultBrushNames.TimeRegion2Hatch, New Long() {1})
            customRegionStart = Start.AddDays(3).AddHours(10)
            AddCustomTimeRegion("Predefined TimeRegion4", customRegionStart, customRegionStart.AddHours(5), DefaultBrushNames.TimeRegion3Hatch)
            SelectedTimeRegion = TimeRegions(0)
        End Sub

        Private Sub AddCustomTimeRegion(ByVal caption As String, ByVal start As Date, ByVal [end] As Date, ByVal brushName As String, ByVal Optional resourceIds As IEnumerable(Of Long) = Nothing)
            Dim res = TimeRegionViewModel.Create()
            res.Caption = caption
            res.Start = start
            res.End = [end]
            res.BrushName = brushName
            res.ResourceIds = resourceIds
            TimeRegions.Add(res)
        End Sub

        Private Sub AddNewTimeRegionInternal(ByVal caption As String)
            Dim res = TimeRegionViewModel.Create()
            res.Caption = caption
            res.Start = regionStart
            res.End = regionStart + regionDuration
            res.BrushName = DefaultBrushNames.TimeRegions(timeRegionNumber Mod DefaultBrushNames.TimeRegions.Length)
            TimeRegions.Add(res)
            timeRegionNumber += 1
            regionStart += regionDuration
            SelectedTimeRegion = res
        End Sub

        Public Sub AddNewTimeRegion(ByVal Optional caption As String = Nothing)
            If String.IsNullOrEmpty(caption) Then caption = "TimeRegion"
            AddNewTimeRegionInternal(caption & timeRegionNumber)
        End Sub

        Public Sub RemoveSelectedTimeRegion()
            TimeRegions.Remove(SelectedTimeRegion)
            SelectedTimeRegion = TimeRegions.FirstOrDefault()
        End Sub
    End Class

    Public Class TimeRegionViewModel

        Public Shared Function Create() As TimeRegionViewModel
            Return ViewModelSource.Create(Function() New TimeRegionViewModel())
        End Function

        Protected Sub New()
        End Sub

        Public Overridable Property Caption As String

        Public Overridable Property Start As Date

        Public Overridable Property [End] As Date

        Public Overridable Property BrushName As String

        Public Overridable ReadOnly Property BorderBrushName As String
            Get
                Return DefaultBrushNames.Resources(DefaultBrushNames.TimeRegions.IndexOf(Function(x) Equals(x, BrushName)) Mod DefaultBrushNames.Resources.Length)
            End Get
        End Property

        Public Overridable Property ResourceIds As IEnumerable(Of Long)
    End Class
End Namespace
